homeassistant/components/broadlink/discovery.py

```diff
@@ -0,0 +1,98 @@
+"""Support for discovering Broadlink devices."""
+import asyncio
+from datetime import timedelta
+from functools import partial
+import logging
+
+import broadlink as blk
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_NAME, CONF_TYPE
+from homeassistant.core import callback
+from homeassistant.helpers import debounce
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import CONF_LOCK, DOMAIN, SUPPORTED_TYPES
+from .helpers import get_broadcast_addrs, get_ip_or_none
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BroadlinkDiscovery:
+    """Manages device discovery."""
+
+    def __init__(self, hass):
+        """Initialize the entity."""
+        self.hass = hass
+        self.timeout = 5
+        self.coordinator = DataUpdateCoordinator(
+            hass,
+            _LOGGER,
+            name="discovery",
+            update_method=self.async_discover,
+            update_interval=timedelta(minutes=10),
+            request_refresh_debouncer=debounce.Debouncer(
+                hass, _LOGGER, cooldown=self.timeout, immediate=True
+            ),
+        )
+        self._unsubscribe = None
+
+    async def async_setup(self):
+        """Set up device discovery."""
+        self._unsubscribe = self.coordinator.async_add_listener(self.update)
+        await self.coordinator.async_refresh()
+
+    async def async_unload(self):
+        """Unload device discovery."""
+        if self._unsubscribe is not None:
+            self._unsubscribe()
+            self._unsubscribe = None
+
+    async def async_discover(self):
+        """Discover Broadlink devices on all available networks."""
+        tasks = [
+            self.hass.async_add_executor_job(
```

Comment:
We don't want to create many executor jobs. Do all related things that needs the thread pool in one executor job.

Reply:
The python-broadlink library is not async, so I thought it would be better to perform these tasks concurrently here with gather(). There will be usually two calls, one for WiFi and the other for Ethernet if the user is running on a RPi. If I move this logic to the library, I think I will need to use multi threads and a queue. I wish didn't have to do so. Ideas?

Reply:
Create a function that calls all calls that should happen in the thread pool. Then schedule that function on the executor.

Reply:
This will multiply the discovery time (5s) by the number of interfaces. Makes no difference for updates (the update interval is 10min), but it will significantly impact the startup time. Here are my suggestions to get around this:
1. Call all discoveries sequentially in a single thread, as you are suggesting, so we avoid bombing the thread loop.
2. Create a separate task to set up device discovery in parallel, so we start the integration fast and discover devices later.
3. Use broadlink.xdiscover() to create flows instantly (this function returns a generator). Do not use an update listener. Schedule a simple task to discover devices and create flows at every 10 minutes. Something like this:
```python3
    async def async_discover(self):
        """Discover Broadlink devices on all available networks."""
        hass = self.hass
        timeout = self.timeout

        def discover():
            broadcast_addrs = hass.data[DOMAIN].config["broadcast_addrs"]
            entries = hass.config_entries.async_entries(DOMAIN)
            hosts_and_macs = {
                (get_ip_or_none(entry.data[CONF_HOST]), entry.data[CONF_MAC])
                for entry in entries
            }
            for addr in broadcast_addrs:
                for device in blk.xdiscover(discover_ip_address=addr, timeout=timeout):
                    host, mac_addr = device.host[0], device.mac.hex()
                    if (host, mac_addr) in hosts_and_macs:
                        continue

                    if device.type not in SUPPORTED_TYPES:
                        continue

                    hass.async_create_task(
                        hass.config_entries.flow.async_init(
                            DOMAIN,
                            context={
                                "source": config_entries.SOURCE_INTEGRATION_DISCOVERY
                            },
                            data={
                                CONF_HOST: device.host[0],
                                CONF_MAC: device.mac.hex(),
                                CONF_TYPE: device.devtype,
                                CONF_NAME: device.name,
                                CONF_LOCK: device.is_locked,
                            },
                        )
                    )

        await hass.async_add_executor_job(discover)
```

What do you think? Can we schedule this task without an update coordinator? We don't need the listener and the event-based features, but the Debouncer is still useful.

Reply:
We can't mix async api inside sync context and vice versa.

Reply:
I finished this part. I still need to adapt the tests to the changes.